File: app/src/main/java/app/di_v/loginactivitybeta/MainActivity.java
package app.di_v.loginactivitybeta;

import android.content.Intent;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.widget.Button;

import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;

/**
 * @author di-v
 */
public class MainActivity extends AppCompatActivity implements View.OnClickListener {

    private static final String TAG = "MainActivity";
    private Button btnLogOut;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        FirebaseUser mAuth = FirebaseAuth.getInstance().getCurrentUser();
        Log.d(TAG, " mAuth: " + mAuth + "; ");

        if (mAuth == null) {
            Log.d(TAG, " start LoginActivity ");
            Intent intent = new Intent(this, LoginActivity.class);
            startActivity(intent);
            this.finish();
        }

        btnLogOut = findViewById(R.id.btn_logout);
        btnLogOut.setOnClickListener(this);
    }

    @Override
    public void onClick(View v) {
        int id = v.getId();

        if (id == R.id.btn_logout) {
            FirebaseAuth.getInstance().signOut();
            Intent intent = new Intent(this, LoginActivity.class);
            startActivity(intent);
            this.finish();
        }
    }
}

File: app/src/main/java/app/di_v/authdemo/ui/main/MainActivity.kt
package app.di_v.authdemo.ui.main

import android.content.Intent
import android.os.Bundle
import androidx.appcompat.app.AppCompatActivity
import app.di_v.authdemo.R
import app.di_v.authdemo.ui.auth.AuthActivity
import com.google.android.gms.auth.api.signin.GoogleSignIn
import com.google.android.gms.auth.api.signin.GoogleSignInOptions
import com.google.firebase.auth.FirebaseAuth
import kotlinx.android.synthetic.main.activity_main.*

class MainActivity: AppCompatActivity() {
    private lateinit var auth: FirebaseAuth
    private val authRequestCode = 7

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        auth = FirebaseAuth.getInstance()

        if (auth.currentUser != null) {
            auth.currentUser!!.email?.let { updateUi(it,  getString(R.string.sign_out)) }
        } else {
            updateUi("need to login","authorization")
        }

        btn_auth_action.setOnClickListener {
            if (auth.currentUser != null) {
                auth.signOut()

                val gso = GoogleSignInOptions.Builder(GoogleSignInOptions.DEFAULT_SIGN_IN)
                    .requestIdToken(getString(R.string.default_web_client_id))
                    .requestEmail()
                    .build()

                val googleSignInClient = GoogleSignIn.getClient(this, gso)

                googleSignInClient.signOut().addOnCompleteListener(this) {
                    updateUi("sign out success!", "authorization")
                }
            } else {
                startActivityForResult(Intent(this, AuthActivity::class.java), authRequestCode)
            }
        }
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        when(requestCode) {
            authRequestCode -> auth.currentUser!!.email?.let { updateUi(it, getString(R.string.sign_out)) }
        }
    }

    private fun updateUi(text: String, textBtn: String) {
        textView.text = text
        btn_auth_action.text = textBtn
    }
}